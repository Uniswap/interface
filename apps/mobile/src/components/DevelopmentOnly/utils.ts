import { useEffect, useState } from 'react'
import { useAppDispatch } from 'src/app/hooks'
import { pushNotification } from 'wallet/src/features/notifications/slice'
import { useActiveAccountAddressWithThrow } from 'wallet/src/features/wallet/hooks'

export const exampleDisconnectedNotification = {
  type: 2,
  // address: '0x...',
  dappName: 'Uniswap Interface',
  event: 1,
  imageUrl: 'https://app.uniswap.org/favicon.png',
  hideDelay: 3000,
}

export const exampleSwapConfirmation = {
  type: 7,
  chainId: 42161,
  hideDelay: 2000,
}

export const exampleSwapSuccess = {
  txStatus: 'failed',
  chainId: 42161,
  txHash: '0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
  // address: '0x...',
  // txId: 'aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa',
  type: 3,
  txType: 'swap',
  inputCurrencyId: '42161-0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
  outputCurrencyId: '42161-0xf97f4df75117a78c1A5a0DBb814Af92458539FB4',
  inputCurrencyAmountRaw: '10000000000000000',
  outputCurrencyAmountRaw: '1356219232855702996',
  tradeType: 0,
  hideDelay: 3000,
}

// easiest to use inside NotificationToastWrapper before any returns
export const useFakeNotification = (ms?: number): void => {
  const [sent, setSent] = useState(false)
  const dispatch = useAppDispatch()
  const activeAddress = useActiveAccountAddressWithThrow()

  useEffect(() => {
    setSent(false)
  }, [ms])

  useEffect(() => {
    if (!sent && activeAddress) {
      dispatch(
        pushNotification({
          ...exampleSwapSuccess,
          hideDelay: ms ?? exampleSwapSuccess.hideDelay,
          address: activeAddress,
        })
      )
      setSent(true)
    }
  }, [activeAddress, dispatch, ms, sent])
}
