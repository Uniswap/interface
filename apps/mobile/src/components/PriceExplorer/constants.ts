import { ElementName } from 'src/features/telemetry/constants'
import { HistoryDuration } from 'wallet/src/data/__generated__/types-and-hooks'
import i18n from 'wallet/src/i18n/i18n'

export const NUM_GRAPHS = 5

export const BUTTON_PADDING = 20

export const CURSOR_INNER_SIZE = 12
export const CURSOR_SIZE = CURSOR_INNER_SIZE + 6
export const LINE_WIDTH = 1

export const TIME_RANGES = [
  [HistoryDuration.Hour, i18n.t('1H'), ElementName.TimeFrame1H],
  [HistoryDuration.Day, i18n.t('1D'), ElementName.TimeFrame1D],
  [HistoryDuration.Week, i18n.t('1W'), ElementName.TimeFrame1W],
  [HistoryDuration.Month, i18n.t('1M'), ElementName.TimeFrame1M],
  [HistoryDuration.Year, i18n.t('1Y'), ElementName.TimeFrame1Y],
] as const
