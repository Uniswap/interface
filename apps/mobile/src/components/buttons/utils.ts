import { withSequence, withSpring, WithSpringConfig } from 'react-native-reanimated'

export function pulseAnimation(
  activeScale: number,
  spingAnimationConfig: WithSpringConfig = { damping: 1, stiffness: 200 }
): number {
  'worklet'
  return withSequence(
    withSpring(activeScale, spingAnimationConfig),
    withSpring(1, spingAnimationConfig)
  )
}
