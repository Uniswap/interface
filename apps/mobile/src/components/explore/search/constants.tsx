export const SEARCH_RESULT_HEADER_KEY = 'header'
