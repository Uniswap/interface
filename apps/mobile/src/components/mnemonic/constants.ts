export const DEFAULT_MNEMONIC_DISPLAY_HEIGHT = 300
export const FULL_MNEMONIC_DISPLAY_HEIGHT = 348
