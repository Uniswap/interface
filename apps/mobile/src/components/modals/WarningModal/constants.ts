import {
  Warning,
  WarningAction,
  WarningLabel,
  WarningSeverity,
} from 'src/components/modals/WarningModal/types'
import WifiIcon from 'ui/src/assets/icons/wifi-slash.svg'
import { AppTFunction } from 'ui/src/i18n/types'

export const getNetworkWarning = (t: AppTFunction): Warning => ({
  type: WarningLabel.NetworkError,
  severity: WarningSeverity.Low,
  action: WarningAction.DisableReview,
  title: t('Youâ€™re offline'),
  icon: WifiIcon,
  message: t(
    'You may have lost internet connection or the network may be down. Please check your internet connection and try again.'
  ),
})
