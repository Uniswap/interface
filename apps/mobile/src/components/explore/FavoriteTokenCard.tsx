import { ImpactFeedbackStyle } from 'expo-haptics'
import React, { memo, useCallback } from 'react'
import { ViewProps } from 'react-native'
import ContextMenu from 'react-native-context-menu-view'
import { FadeIn, FadeOut } from 'react-native-reanimated'
import { useAppDispatch } from 'src/app/hooks'
import { useExploreTokenContextMenu } from 'src/components/explore/hooks'
import RemoveButton from 'src/components/explore/RemoveButton'
import { Loader } from 'src/components/loading'
import { useTokenDetailsNavigation } from 'src/components/TokenDetails/hooks'
import { SectionName } from 'src/features/telemetry/constants'
import { disableOnPress } from 'src/utils/disableOnPress'
import { usePollOnFocusOnly } from 'src/utils/hooks'
import { AnimatedTouchableArea, Flex, Text } from 'ui/src'
import { borderRadii, imageSizes } from 'ui/src/theme'
import { NumberType } from 'utilities/src/format/types'
import { BaseCard } from 'wallet/src/components/BaseCard/BaseCard'
import { TokenLogo } from 'wallet/src/components/CurrencyLogo/TokenLogo'
import { RelativeChange } from 'wallet/src/components/text/RelativeChange'
import { ChainId } from 'wallet/src/constants/chains'
import { PollingInterval } from 'wallet/src/constants/misc'
import { isNonPollingRequestInFlight } from 'wallet/src/data/utils'
import { useFavoriteTokenCardQuery } from 'wallet/src/data/__generated__/types-and-hooks'
import { fromGraphQLChain } from 'wallet/src/features/chains/utils'
import { currencyIdToContractInput } from 'wallet/src/features/dataApi/utils'
import { removeFavoriteToken } from 'wallet/src/features/favorites/slice'
import { useLocalizationContext } from 'wallet/src/features/language/LocalizationContext'
import { getSymbolDisplayText } from 'wallet/src/utils/currency'

export const FAVORITE_TOKEN_CARD_LOADER_HEIGHT = 114

type FavoriteTokenCardProps = {
  currencyId: string
  isEditing?: boolean
  setIsEditing: (update: boolean) => void
} & ViewProps

function FavoriteTokenCard({
  currencyId,
  isEditing,
  setIsEditing,
  ...rest
}: FavoriteTokenCardProps): JSX.Element {
  const dispatch = useAppDispatch()
  const tokenDetailsNavigation = useTokenDetailsNavigation()
  const { convertFiatAmountFormatted } = useLocalizationContext()

  const { data, networkStatus, startPolling, stopPolling } = useFavoriteTokenCardQuery({
    variables: currencyIdToContractInput(currencyId),
    // Rely on cache for fast favoriting UX, and poll for updates.
    fetchPolicy: 'cache-first',
    returnPartialData: true,
  })

  usePollOnFocusOnly(startPolling, stopPolling, PollingInterval.Fast)

  const token = data?.token

  // Mirror behavior in top tokens list, use first chain the token is on for the symbol
  const chainId = fromGraphQLChain(token?.chain) ?? ChainId.Mainnet

  const price = convertFiatAmountFormatted(
    token?.project?.markets?.[0]?.price?.value,
    NumberType.FiatTokenPrice
  )
  const pricePercentChange = token?.project?.markets?.[0]?.pricePercentChange24h?.value

  const onRemove = useCallback(() => {
    if (currencyId) {
      dispatch(removeFavoriteToken({ currencyId }))
    }
  }, [currencyId, dispatch])

  const onEditFavorites = useCallback(() => {
    setIsEditing(true)
  }, [setIsEditing])

  const { menuActions, onContextMenuPress } = useExploreTokenContextMenu({
    chainId,
    currencyId,
    analyticsSection: SectionName.ExploreFavoriteTokensSection,
    onEditFavorites,
  })

  const onPress = (): void => {
    if (isEditing || !currencyId) return
    tokenDetailsNavigation.preload(currencyId)
    tokenDetailsNavigation.navigate(currencyId)
  }

  if (isNonPollingRequestInFlight(networkStatus)) {
    return <Loader.Favorite height={FAVORITE_TOKEN_CARD_LOADER_HEIGHT} />
  }

  return (
    <ContextMenu
      actions={menuActions}
      disabled={isEditing}
      style={{ borderRadius: borderRadii.rounded16 }}
      onPress={onContextMenuPress}
      {...rest}>
      <AnimatedTouchableArea
        hapticFeedback
        borderRadius="$rounded16"
        entering={FadeIn}
        exiting={FadeOut}
        hapticStyle={ImpactFeedbackStyle.Light}
        m="$spacing4"
        testID={`token-box-${token?.symbol}`}
        onLongPress={disableOnPress}
        onPress={onPress}>
        <BaseCard.Shadow>
          <Flex alignItems="flex-start" gap="$spacing8">
            <Flex row gap="$spacing4" justifyContent="space-between">
              <Flex grow row alignItems="center" gap="$spacing8">
                <TokenLogo
                  chainId={chainId ?? undefined}
                  size={imageSizes.image20}
                  symbol={token?.symbol ?? undefined}
                  url={token?.project?.logoUrl ?? undefined}
                />
                <Text variant="body1">{getSymbolDisplayText(token?.symbol)}</Text>
              </Flex>
              {isEditing ? (
                <RemoveButton onPress={onRemove} />
              ) : (
                <Flex height={imageSizes.image24} />
              )}
            </Flex>
            <Flex gap="$spacing2">
              <Text adjustsFontSizeToFit numberOfLines={1} variant="heading3">
                {price}
              </Text>
              <RelativeChange
                arrowSize="$icon.16"
                change={pricePercentChange ?? undefined}
                semanticColor={true}
                variant="subheading2"
              />
            </Flex>
          </Flex>
        </BaseCard.Shadow>
      </AnimatedTouchableArea>
    </ContextMenu>
  )
}

export default memo(FavoriteTokenCard)
