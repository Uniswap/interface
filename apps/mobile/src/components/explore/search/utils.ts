import { searchResultId } from 'src/features/explore/searchHistorySlice'
import {
  NFTCollectionSearchResult,
  SearchResultType,
  TokenSearchResult,
} from 'src/features/explore/SearchResult'
import { Chain, ExploreSearchQuery } from 'wallet/src/data/__generated__/types-and-hooks'
import { fromGraphQLChain } from 'wallet/src/features/chains/utils'
import { SEARCH_RESULT_HEADER_KEY } from './constants'
import { SearchResultOrHeader } from './types'

const MAX_TOKEN_RESULTS_COUNT = 4

type ExploreSearchResult = NonNullable<ExploreSearchQuery>

// Formats the tokens portion of explore search results into sorted array of TokenSearchResult
export function formatTokenSearchResults(
  data: ExploreSearchResult['searchTokens'],
  searchQuery: string
): TokenSearchResult[] | undefined {
  if (!data) return

  // Prevent showing "duplicate" token search results for tokens that are on multiple chains
  // and share the same TokenProject id. Only show the token that has the highest 1Y Uniswap trading volume
  // ex. UNI on Mainnet, Arbitrum, Optimism -> only show UNI on Mainnet b/c it has highest 1Y volume
  const tokenResultsMap = data.reduce<Record<string, TokenSearchResult & { volume1D: number }>>(
    (tokensMap, token) => {
      if (!token) return tokensMap

      const { chain, address, symbol, project, market } = token
      const chainId = fromGraphQLChain(chain)

      if (!chainId || !project) return tokensMap

      const { name, safetyLevel, logoUrl } = project

      const tokenResult: TokenSearchResult & { volume1D: number } = {
        type: SearchResultType.Token,
        chainId,
        address: address ?? null,
        name: name ?? null,
        symbol: symbol ?? '',
        safetyLevel: safetyLevel ?? null,
        logoUrl: logoUrl ?? null,
        volume1D: market?.volume?.value ?? 0,
      }

      // For token results that share the same TokenProject id, use the token with highest volume
      const currentTokenResult = tokensMap[project.id]
      if (!currentTokenResult || tokenResult.volume1D > currentTokenResult.volume1D) {
        tokensMap[project.id] = tokenResult
      }
      return tokensMap
    },
    {}
  )

  return Object.values(tokenResultsMap)
    .slice(0, MAX_TOKEN_RESULTS_COUNT)
    .sort((res1: TokenSearchResult, res2: TokenSearchResult) => {
      const res1Match = isExactTokenSearchResultMatch(res1, searchQuery)
      const res2Match = isExactTokenSearchResultMatch(res2, searchQuery)

      if (res1Match && !res2Match) {
        return -1
      } else if (!res1Match && res2Match) {
        return 1
      } else {
        return 0
      }
    })
}

function isExactTokenSearchResultMatch(searchResult: TokenSearchResult, query: string): boolean {
  return (
    searchResult.name?.toLowerCase() === query.toLowerCase() ||
    searchResult.symbol.toLowerCase() === query.toLowerCase()
  )
}

export function formatNFTCollectionSearchResults(
  data: ExploreSearchResult['nftCollections']
): NFTCollectionSearchResult[] | undefined {
  if (!data) return

  return data.edges.reduce<NFTCollectionSearchResult[]>((accum, { node }) => {
    const formatted = gqlNFTToNFTCollectionSearchResult(node)
    if (formatted) {
      accum.push(formatted)
    }
    return accum
  }, [])
}

type NFTCollectionItemResult = NonNullable<
  NonNullable<NonNullable<NonNullable<ExploreSearchResult['nftCollections']>>['edges']>[0]
>['node']

export const gqlNFTToNFTCollectionSearchResult = (
  node: NFTCollectionItemResult
): NFTCollectionSearchResult | null => {
  const contract = node?.nftContracts?.[0]
  // Only show NFT results that have fully populated results
  const chainId = fromGraphQLChain(node?.nftContracts?.[0]?.chain ?? Chain.Ethereum)
  if (node.name && contract?.address && chainId) {
    return {
      type: SearchResultType.NFTCollection,
      chainId,
      address: contract.address,
      name: node.name,
      imageUrl: node?.image?.url ?? null,
      isVerified: Boolean(node.isVerified),
    }
  }
  return null
}

export const getSearchResultId = (searchResult: SearchResultOrHeader): string => {
  if (searchResult.type === SEARCH_RESULT_HEADER_KEY) {
    return searchResult.title
  }
  // Unique ID for each search result
  return searchResultId(searchResult)
}
