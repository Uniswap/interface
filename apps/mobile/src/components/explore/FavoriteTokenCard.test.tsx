import { makeMutable } from 'react-native-reanimated'
import configureMockStore from 'redux-mock-store'
import FavoriteTokenCard, { FavoriteTokenCardProps } from 'src/components/explore/FavoriteTokenCard'
import { act, cleanup, fireEvent, render, waitFor } from 'src/test/test-utils'
import { FiatCurrency } from 'uniswap/src/features/fiatCurrency/constants'
import { Language } from 'uniswap/src/features/language/constants'
import {
  ON_PRESS_EVENT_PAYLOAD,
  SAMPLE_CURRENCY_ID_1,
  amount,
  ethToken,
  tokenMarket,
  tokenProject,
} from 'uniswap/src/test/fixtures'
import { queryResolvers } from 'uniswap/src/test/utils'
import { getSymbolDisplayText } from 'uniswap/src/utils/currency'

const mockedNavigation = {
  navigate: jest.fn(),
}

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native')
  return {
    ...actualNav,
    // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
    useNavigation: () => mockedNavigation,
  }
})

const mockStore = configureMockStore()

const favoriteToken = ethToken({
  project: tokenProject(),
  market: tokenMarket({
    price: amount({ value: 12345.67 }),
    pricePercentChange: amount({ value: 4.56 }),
  }),
})

const touchableId = `token-box-${favoriteToken.symbol}`

const defaultProps: FavoriteTokenCardProps = {
  currencyId: SAMPLE_CURRENCY_ID_1,
  pressProgress: makeMutable(0),
  dragActivationProgress: makeMutable(0),
  setIsEditing: jest.fn(),
  isEditing: false,
}

const { resolvers } = queryResolvers({
  token: () => favoriteToken,
})

describe('FavoriteTokenCard', () => {
  it('renders without error', async () => {
    const tree = render(<FavoriteTokenCard {...defaultProps} />)

    expect(tree).toMatchSnapshot()
    cleanup()
  })

  describe('when token data is being fetched', () => {
    it('renders loader', async () => {
      const { queryByTestId } = render(<FavoriteTokenCard {...defaultProps} />, { resolvers })

      const loader = queryByTestId('loader/favorite')

      // loading
      expect(loader).toBeTruthy()

      // loading finished
      await waitFor(() => {
        expect(queryByTestId(touchableId)).toBeTruthy()
      })
    })
  })

  describe('when token data is available', () => {
    const cases = [
      { test: 'symbol', value: getSymbolDisplayText(favoriteToken.symbol)! },
      { test: 'price', value: '$12,345.67' },
      { test: 'relative price change', value: '4.56%' },
    ]

    it.each(cases)('renders correct $test', async ({ value }) => {
      const { queryByText } = render(<FavoriteTokenCard {...defaultProps} />, { resolvers })

      await waitFor(() => {
        expect(queryByText(value)).toBeTruthy()
      })
    })

    it('navigates to the token details screen when pressed', async () => {
      const { findByTestId } = render(<FavoriteTokenCard {...defaultProps} />, { resolvers })

      const touchable = await findByTestId(`token-box-${favoriteToken.symbol}`)
      act(() => {
        fireEvent.press(touchable, ON_PRESS_EVENT_PAYLOAD)
      })

      expect(mockedNavigation.navigate).toHaveBeenCalledTimes(1)
      expect(mockedNavigation.navigate).toHaveBeenCalledWith('TokenDetails', {
        currencyId: SAMPLE_CURRENCY_ID_1, // passed in component props
      })
    })

    it('does not show remove button when not in edit mode', async () => {
      const { findByTestId } = render(<FavoriteTokenCard {...defaultProps} />, { resolvers })

      const removeButton = await findByTestId('explore/remove-button')

      await waitFor(() => {
        expect(removeButton).toHaveAnimatedStyle({ opacity: 0 })
      })
    })
  })

  describe('edit mode', () => {
    it('shows remove button when in edit mode', async () => {
      const { findByTestId } = render(<FavoriteTokenCard {...defaultProps} isEditing />, {
        resolvers,
      })

      const removeButton = await findByTestId('explore/remove-button')

      await waitFor(() => {
        expect(removeButton).toHaveAnimatedStyle({ opacity: 1 })
      })
    })

    it('dispatches removeFavoriteToken action when remove button is pressed', async () => {
      const store = mockStore({
        favorites: { tokens: [] },
        userSettings: { currentCurrency: FiatCurrency.UnitedStatesDollar, currentLanguage: Language.English },
      })
      const { findByTestId } = render(<FavoriteTokenCard {...defaultProps} isEditing />, {
        resolvers,
        store,
      })

      const removeButton = await findByTestId('explore/remove-button')
      await act(() => {
        fireEvent.press(removeButton, ON_PRESS_EVENT_PAYLOAD)
      })

      const actions = store.getActions()
      expect(actions).toEqual([
        { type: 'favorites/removeFavoriteToken', payload: { currencyId: SAMPLE_CURRENCY_ID_1 } },
      ])
    })
  })
})
