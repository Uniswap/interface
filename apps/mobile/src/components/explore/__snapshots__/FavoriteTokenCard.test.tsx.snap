// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteTokenCard renders without error 1`] = `
<View
  actions={
    [
      {
        "onPress": [Function],
        "systemIcon": "heart",
        "title": "Favorite token",
      },
      {
        "onPress": [Function],
        "systemIcon": "square.and.pencil",
        "title": "Edit favorites",
      },
      {
        "onPress": [Function],
        "systemIcon": "arrow.2.squarepath",
        "title": "Swap",
      },
      {
        "onPress": [Function],
        "systemIcon": "qrcode",
        "title": "Receive",
      },
    ]
  }
  disabled={false}
  onPress={[Function]}
  style={
    {
      "borderRadius": 16,
    }
  }
>
  <View
    collapsable={false}
    hitSlop={
      {
        "bottom": 5,
        "left": 5,
        "right": 5,
        "top": 5,
      }
    }
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": {
          "dynamic": {
            "dark": "#131313",
            "light": "#FFFFFF",
          },
        },
        "borderBottomColor": "rgba(34,34,34,0.05)",
        "borderBottomLeftRadius": 16,
        "borderBottomRightRadius": 16,
        "borderBottomWidth": 1,
        "borderLeftColor": "rgba(34,34,34,0.05)",
        "borderLeftWidth": 1,
        "borderRightColor": "rgba(34,34,34,0.05)",
        "borderRightWidth": 1,
        "borderStyle": "solid",
        "borderTopColor": "rgba(34,34,34,0.05)",
        "borderTopLeftRadius": 16,
        "borderTopRightRadius": 16,
        "borderTopWidth": 1,
        "flexDirection": "column",
        "opacity": 1,
        "overflow": "hidden",
        "shadowColor": "rgb(0,0,0)",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.0196078431372549,
        "shadowRadius": 6,
        "transform": [
          {
            "scale": 1,
          },
        ],
      }
    }
    testID="token-box-undefined"
  >
    <View
      style={
        {
          "alignItems": "flex-start",
          "flexDirection": "column",
          "gap": 8,
          "paddingBottom": 12,
          "paddingLeft": 12,
          "paddingRight": 12,
          "paddingTop": 12,
        }
      }
    >
      <View
        style={
          {
            "flexDirection": "row",
            "gap": 4,
            "justifyContent": "space-between",
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "flexGrow": 1,
              "gap": 8,
            }
          }
        >
          <View
            onLayout={[Function]}
            style={
              {
                "flexDirection": "column",
                "opacity": 0,
              }
            }
            testID="shimmer-placeholder"
          >
            <View
              style={
                {
                  "flexDirection": "column",
                }
              }
            >
              <View
                style={
                  {
                    "backgroundColor": {
                      "dynamic": {
                        "dark": "rgba(255, 255, 255, 0.38)",
                        "light": "rgba(19, 19, 19, 0.35)",
                      },
                    },
                    "borderBottomLeftRadius": 999999,
                    "borderBottomRightRadius": 999999,
                    "borderTopLeftRadius": 999999,
                    "borderTopRightRadius": 999999,
                    "flexDirection": "column",
                    "height": 20,
                    "width": 20,
                  }
                }
              />
            </View>
          </View>
          <Text
            allowFontScaling={true}
            maxFontSizeMultiplier={1.4}
            style={
              {
                "color": {
                  "dynamic": {
                    "dark": "#FFFFFF",
                    "light": "#131313",
                  },
                },
                "fontFamily": "Basel Grotesk",
                "fontSize": 19,
                "fontWeight": "400",
                "lineHeight": 24.7,
              }
            }
            suppressHighlighting={true}
          />
        </View>
        <View
          collapsable={false}
          disabled={true}
          hitSlop={
            {
              "bottom": 5,
              "left": 5,
              "right": 5,
              "top": 5,
            }
          }
          style={
            {
              "alignItems": "center",
              "backgroundColor": {
                "dynamic": {
                  "dark": "rgba(255, 255, 255, 0.38)",
                  "light": "rgba(19, 19, 19, 0.35)",
                },
              },
              "borderBottomLeftRadius": 999999,
              "borderBottomRightRadius": 999999,
              "borderTopLeftRadius": 999999,
              "borderTopRightRadius": 999999,
              "flexDirection": "column",
              "height": 24,
              "justifyContent": "center",
              "opacity": 0,
              "transform": [
                {
                  "scale": 1,
                },
              ],
              "width": 24,
              "zIndex": 1080,
            }
          }
          testID="explore/remove-button"
        >
          <View
            style={
              {
                "backgroundColor": {
                  "dynamic": {
                    "dark": "#131313",
                    "light": "#FFFFFF",
                  },
                },
                "borderBottomLeftRadius": 12,
                "borderBottomRightRadius": 12,
                "borderTopLeftRadius": 12,
                "borderTopRightRadius": 12,
                "flexDirection": "column",
                "height": 2,
                "width": 10,
              }
            }
          />
        </View>
      </View>
      <View
        style={
          {
            "flexDirection": "column",
            "gap": 2,
          }
        }
      >
        <View
          onLayout={[Function]}
          style={
            {
              "flexDirection": "column",
              "opacity": 0,
            }
          }
          testID="shimmer-placeholder"
        >
          <View
            style={
              {
                "flexDirection": "column",
              }
            }
          >
            <View
              style={
                {
                  "backgroundColor": {
                    "dynamic": {
                      "dark": "rgba(255, 255, 255, 0.38)",
                      "light": "rgba(19, 19, 19, 0.35)",
                    },
                  },
                  "borderBottomLeftRadius": 12,
                  "borderBottomRightRadius": 12,
                  "borderTopLeftRadius": 12,
                  "borderTopRightRadius": 12,
                  "flexDirection": "column",
                  "height": 30,
                  "width": 90,
                }
              }
              testID="loader/favorite/price"
            />
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={
            {
              "flexDirection": "column",
              "opacity": 0,
            }
          }
          testID="shimmer-placeholder"
        >
          <View
            style={
              {
                "flexDirection": "column",
              }
            }
          >
            <View
              style={
                {
                  "backgroundColor": {
                    "dynamic": {
                      "dark": "rgba(255, 255, 255, 0.38)",
                      "light": "rgba(19, 19, 19, 0.35)",
                    },
                  },
                  "borderBottomLeftRadius": 12,
                  "borderBottomRightRadius": 12,
                  "borderTopLeftRadius": 12,
                  "borderTopRightRadius": 12,
                  "flexDirection": "column",
                  "height": 20,
                  "width": 60,
                }
              }
              testID="loader/favorite/priceChange"
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
